validator_worker/src/follower.rs

```diff
@@ -1,12 +1,139 @@
+use primitives::BalancesMap;
 
-impl Validator for Follower {
-    fn tick(&self, _channel: Channel) -> ValidatorFuture<()> {
-        futures::future::ok(()).boxed()
+use crate::core::follower_rules::{is_healthy, is_valid_transition};
+use crate::heartbeat::heartbeat;
+use crate::sentry_interface::SentryApi;
+use crate::{get_state_root_hash, producer};
+use chrono::Utc;
+
+enum InvalidNewState {
+    RootHash,
+    Signature,
+    Transition,
+}
+
+enum NewStateResult {
+    Ok,
+    Err(InvalidNewState),
+}
+
+pub async fn tick<A: Adapter + 'static>(iface: &SentryApi<A>) -> Result<(), Box<dyn Error>> {
+    let from = iface.channel.spec.validators.leader().id.clone();
+    let new_msg_response = await!(iface.get_latest_msg(from, "NewState".to_string()))?;
+    let new_msg = new_msg_response
+        .msg
+        .get(0)
+        .and_then(|message_types| match message_types {
+            MessageTypes::NewState(new_state) => Some(new_state.clone()),
+            _ => None,
+        });
+    let our_latest_msg_response =
+        await!(iface.get_our_latest_msg("ApproveState+RejectState".to_string()))?;
+    let our_latest_msg_state_root = our_latest_msg_response
+        .msg
+        .get(0)
+        .and_then(|message_types| match message_types {
+            MessageTypes::ApproveState(approve_state) => Some(approve_state.state_root.clone()),
+            MessageTypes::RejectState(reject_state) => Some(reject_state.state_root.clone()),
+            _ => None,
+        });
+
+    let latest_is_responded_to = match (&new_msg, &our_latest_msg_state_root) {
+        (Some(new_msg), Some(state_root)) => &new_msg.state_root == state_root,
+        (_, _) => false,
+    };
+
+    let (balances, _) = await!(producer::tick(&iface))?;
+
+    if let (Some(new_state), false) = (new_msg, latest_is_responded_to) {
+        await!(on_new_state(&iface, &balances, &new_state))?;
+    }
+
+    await!(heartbeat(&iface, balances)).map(|_| ())
+}
+
+async fn on_new_state<'a, A: Adapter + 'static>(
+    iface: &'a SentryApi<A>,
+    balances: &'a BalancesMap,
+    new_state: &'a NewState,
+) -> Result<NewStateResult, Box<dyn Error>> {
+    let proposed_balances = new_state.balances.clone();
+    let proposed_state_root = new_state.state_root.clone();
+
+    if proposed_state_root != hex::encode(get_state_root_hash(&iface, &proposed_balances)?) {
+        return Ok(await!(on_error(
+            &iface,
+            &new_state,
+            InvalidNewState::RootHash
+        )));
     }
+
+    if !iface.adapter.verify(
+        &iface.channel.spec.validators.leader().id,
+        &proposed_state_root,
+        &new_state.signature,
+    )? {
+        return Ok(await!(on_error(
+            &iface,
+            &new_state,
+            InvalidNewState::Signature
+        )));
+    }
+
+    let last_approve_response = await!(iface.get_last_approved())?;
+    // TODO: Check if it's possible to have an empty response (i.e. we don't have LastApprove)
```

Comment:
See https://github.com/AdExNetwork/adex-validator/blob/master/services/validatorWorker/follower.js#L45

Reply:
last_approved can have an empty response.
It's a JS check against "{}" so in the case of Rust it would be None

Reply:
So we have:
```
pub struct LastApprovedResponse {
    pub last_approved: Option<LastApproved>, <---- This?
    pub heartbeats: Option<Vec<Heartbeat>>,
}
```

And do we have this?

```
pub struct LastApproved {
    pub new_state: Option<NewState>, <--- Option here
    /// ApproveState can be None if the channel is brand new and we have only NewState atm
    pub approved_state: Option<ApproveState>, <--- And as we discussed, this is an edge case so we do have this
}
```

Reply:
Yes this is correct.